cmd/erasure-healfile.go

```diff
@@ -16,72 +16,136 @@
+	algorithm BitrotAlgorithm, checksums [][]byte) (f ErasureFileInfo,
+	err error) {
 
-// HealFile tries to reconstruct a bitrot encoded file spread over all available disks. HealFile will read the valid parts of the file,
-// reconstruct the missing data and write the reconstructed parts back to the disks.
-// It will try to read the valid parts from the file under the given volume and path and tries to reconstruct the file under the given
-// healVolume and healPath. The given algorithm will be used to verify the valid parts and to protect the reconstructed file.
-func (s ErasureStorage) HealFile(offlineDisks []StorageAPI, volume, path string, blocksize int64, healVolume, healPath string, size int64, algorithm BitrotAlgorithm, checksums [][]byte) (f ErasureFileInfo, err error) {
 	if !algorithm.Available() {
 		return f, traceError(errBitrotHashAlgoInvalid)
 	}
+
+	// Initialization
 	f.Checksums = make([][]byte, len(s.disks))
-	hashers, verifiers := make([]hash.Hash, len(s.disks)), make([]*BitrotVerifier, len(s.disks))
+	hashers := make([]hash.Hash, len(s.disks))
+	verifiers := make([]*BitrotVerifier, len(s.disks))
 	for i, disk := range s.disks {
-		if disk == OfflineDisk {
+		switch {
+		case staleDisks[i] != nil:
 			hashers[i] = algorithm.New()
-		} else {
+		case disk == OfflineDisk:
+			// no verifier or hasher needed for an offline
+			// disk
+			continue
+		default:
 			verifiers[i] = NewBitrotVerifier(algorithm, checksums[i])
 			f.Checksums[i] = checksums[i]
 		}
 	}
-	blocks := make([][]byte, len(s.disks))
+
+	// Scan part files in disk, block-by-block reconstruct it and
+	// write to stale disks.
 	chunksize := getChunkSize(blocksize, s.dataBlocks)
-	for offset := int64(0); offset < size; offset += blocksize {
-		if size < blocksize {
-			blocksize = size
+	var chunkOffset, blockOffset int64
+	for ; blockOffset < size; blockOffset += blocksize {
+		// last iteration may have less than blocksize data
+		// left, so chunksize needs to be recomputed.
+		if size < blockOffset+blocksize {
+			blocksize = size - blockOffset
 			chunksize = getChunkSize(blocksize, s.dataBlocks)
 		}
+
+		// read a chunk from each disk, until we have
+		// `s.dataBlocks` number of chunks set to non-nil in
+		// `blocks`
+		blocks := make([][]byte, len(s.disks))
+		var buffer []byte
 		numReads := 0
 		for i, disk := range s.disks {
-			if disk != OfflineDisk {
-				if blocks[i] == nil {
-					blocks[i] = make([]byte, chunksize)
-				}
-				blocks[i] = blocks[i][:chunksize]
-				if !verifiers[i].IsVerified() {
-					_, err = disk.ReadFileWithVerify(volume, path, offset, blocks[i], verifiers[i])
-				} else {
-					_, err = disk.ReadFile(volume, path, offset, blocks[i])
-				}
-				if err != nil {
-					blocks[i] = nil
-				} else {
-					numReads++
-				}
-				if numReads == s.dataBlocks { // we have enough data to reconstruct
-					break
-				}
+			// skip reading from offline or stale disks
+			if disk == OfflineDisk || staleDisks[i] != nil {
```

Comment:
Thinking more about this, `disk == OfflineDisk` is incorrect, because if `disk` is `nil` it can also mean that the disk is online but outdated.

So the comparison is `disk == nil` will be correct.

Also you don't need `staleDisks[i] != nil` because if `disk` is not `nil` it already means that it is not outdated.

You could use the same term `outDatedDisks` that is used in the caller instead of `staleDisks` to increase readability.

Reply:
@krishnasrinivas 
This loop is over the `s.disks` slice. This is a list of disks built before any verification of data on the disks based on the erasure ordering determined from `xl.json`. The caller ensures that a majority of this slice is non-nil. There is nothing wrong with considering the nil disks in this slice as offline (also in the code `OfflineDisk == nil` - I am not sure if you knew). Changing `OfflineDisk` to `nil` or vice-versa has no impact on the correctness of the algorithm.

`staleDisks` as defined in the function's doc (at the beginning) is a slice with non-nil items being online disks with stale (or no) data. These are the only disks on which healing is attempted here.

The check `staleDisks[i] != nil` is useful, because we need not try to read a valid shard (for reconstruction) from a disk that is confirmed to be stale (the caller checks that the disk is outdated). We only get a verification error.

In addition, if `staleDisks[i] != nil`, then `disk != nil` as staleDisks is a subset of `s.disks`, however `disk != nil` has no implication on the i-th staleDisk being nil or not.

Now, with this clarification, I think the algorithm implemented in this function should be clear - to me it looks correct.

Reply:
> This loop is over the s.disks slice. This is a list of disks built before any verification of data on the disks based on the erasure ordering determined from xl.json. The caller ensures that a majority of this slice is non-nil.

I think you are assuming that `s.disks` contains all the online disks but just ordered correctly.
Check out `listOnlineDisks()` implementation, it returns the latest disks based on common modTime. Perhaps we can rename this function to `listLatestDisks()` as it does not list Online disks which are outdated.

> Changing OfflineDisk to nil or vice-versa has no impact on the correctness of the algorithm.

I was pointing to the readability because `if disk == OfflineDisk` reads as `if the disk is Offline` which is incorrect as the particular disk can be online but outdated.

> In addition, if staleDisks[i] != nil, then disk != nil as staleDisks is a subset of s.disks

Incorrect, `s.disks` and `staleDisks` are actually mutually exclusive *except* for the disks that are offline.

> however disk != nil has no implication on the i-th staleDisk being nil or not

if `s.disk[i] != nil` it definitely implies that `staleDisk[i] == nil` however if `s.disks[i] == nil` it does not imply that `staleDisk[i] != nil` as that disk may be offline.

Reply:
I am starting over the clarification in the main thread. See below.